src/Microsoft.ML.Data/Transforms/TermTransform.cs

```diff
@@ -674,85 +720,194 @@ public override void Save(ModelSaveContext ctx)
                     Host.CheckValue(c, nameof(ctx));
                     c.CheckAtModel();
                     c.SetVersionInfo(GetTermManagerVersionInfo());
-                    c.Writer.Write(_termMap.Length);
-                    foreach (var term in _termMap)
-                        term.Map.Save(c, this);
+                    c.Writer.Write(_unboundMaps.Length);
+                    foreach (var term in _unboundMaps)
+                        term.Save(c, Host, CodecFactory);
 
                     c.SaveTextStream("Terms.txt",
                         writer =>
                         {
-                            foreach (var map in _termMap)
+                            foreach (var map in _unboundMaps)
                                 map.WriteTextTerms(writer);
                         });
                 });
         }
 
-        protected override JToken SaveAsPfaCore(BoundPfaContext ctx, int iinfo, ColInfo info, JToken srcToken)
+        protected override IRowMapper MakeRowMapper(ISchema schema)
+          => new Mapper(this, schema);
+
+        protected override void CheckInputColumn(ISchema inputSchema, int col, int srcCol)
+        {
+            if ((inputSchema.GetColumnType(srcCol).ItemType.RawKind == default))
+                throw Host.ExceptSchemaMismatch(nameof(inputSchema), "input", ColumnPairs[col].input, "image", inputSchema.GetColumnType(srcCol).ToString());
+        }
+
+        private sealed class Mapper : MapperBase, ISaveAsOnnx, ISaveAsPfa
         {
-            Contracts.AssertValue(ctx);
-            Contracts.Assert(0 <= iinfo && iinfo < Infos.Length);
-            Contracts.Assert(Infos[iinfo] == info);
-            Contracts.AssertValue(srcToken);
-            Contracts.Assert(CanSavePfa);
+            private readonly ColumnType[] _types;
+            private readonly TermTransform _parent;
+            private readonly ColInfo[] _infos;
```

Comment:
>_infos [](start = 39, length = 6)

can you zip `_types` into `_infos`?

Reply:
I mostly use ColInfo to pass information about schema to TermMap builders. So I can go scan data and build dictionaries.
_types is output type of the column, and depends on TermMap (it's gonna be Key of U4, but you don't know cardinality of it until you build map.

I can put them together, but they would be somewhat weird.


---
In reply to: [214480038](https://github.com/dotnet/machinelearning/pull/759#discussion_r214480038) [](ancestors = 214480038)

Reply:
I see

---
In reply to: [214482092](https://github.com/dotnet/machinelearning/pull/759#discussion_r214482092) [](ancestors = 214482092,214480038)